### PR TITLE
IDEX-3087: added description to Recipe and predictable recipe loader

### PR DESCRIPTION
**Pull request by akorneta**

### CODE REVIEW

File: platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/recipe/RecipeLoader.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        this.recipesName = recipesName;
+        this.recipeDao = recipeDao;
+        this.gson = new Gson();
+    }
+
+    @PostConstruct
+    public void start() throws ServerException {
+        URL recipeUrl = Thread.currentThread().getContextClassLoader().getResource(recipesName);
+        if (recipeUrl == null) {
+            throw new ServerException("Impossible get predefined recipes");
```
Comment (evoevodin) [inline platform-api/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/recipe/RecipeLoader.java]: I think this message should be more informative 
i.e. _File with predefined recipes '/path/to/recipes.json' doesn't exist_

File: platform-api/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/recipe/RecipeLoaderTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @BeforeMethod
+    public void setUp() throws Exception {
+        recipeLoader = spy(new RecipeLoader("recipes.json", recipeDao));
+    }
+
+    @Test
+    public void shouldLoadPredictableRecipesFromValidJson() throws Exception {
+        doNothing().when(recipeDao).create(any());
+
+        recipeLoader.start();
```
Comment (mshaposhnik) [inline platform-api/che-core-api-machine/src/test/java/org/eclipse/che/api/machine/server/recipe/RecipeLoaderTest.java]: It's useful here to check that some required action performed, for example dao.create() is called

**Comment by evoevodin** [comment]
Flow which i see:
<code>RecipeLoader</code> loads recipes using <code>WorkspaceDao</code>
<code>WorkspaceDao</code> stores recipes in memory
On shutdown <code>WorkspaceDao</code> saves recipes to _FS_ 
On _Che_ start <code>WorkspaceDao</code> loads recipes which were saved on _FS_.

So i think there is conflict between <code>RecipeLoader</code> & <code>WorkspaceDao</code>

File: platform-api/che-core-api-infrastructure-local/src/test/java/org/eclipse/che/api/local/PredefinedRecipeLoaderTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Test
+    public void shouldLoadPredictableRecipesFromValidJson() throws Exception {
+        recipeLoader = new PredefinedRecipeLoader("recipes.json", recipeDao);
+        List<RecipeImpl> result = new ArrayList<>();
+        Mockito.doAnswer(r -> result.add((RecipeImpl)r.getArguments()[0])).when(recipeDao).create(any());
+
+        recipeLoader.start();
+
+        Assert.assertEquals(result.size(), 2);
```
Comment (evoevodin) [inline platform-api/che-core-api-infrastructure-local/src/test/java/org/eclipse/che/api/local/PredefinedRecipeLoaderTest.java]: I think in this case is better to use <code>verify</code> + <code>times()</code>

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalRecipeDaoImpl.java [hunk 4]
```diff
@@ -65,24 +65,18 @@ public LocalRecipeDaoImpl(LocalStorageFactory storageFactory) throws IOException
 
     @Inject
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalRecipeDaoImpl.java]: Do you need it now?

Comment (akorneta) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalRecipeDaoImpl.java]: +  it's useless

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/PredefinedRecipeLoader.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    private static final Gson GSON = new GsonBuilder().registerTypeAdapter(Permissions.class, new PermissionsSerializer())
+                                                      .registerTypeAdapter(Group.class, new GroupSerializer())
+                                                      .create();
+
+    private final String    recipesFileName;
+    private final RecipeDao recipeDao;
+
+    @Inject
+    public PredefinedRecipeLoader(@Named("che.local.infrastructure.recipes.filename") String recipesName, RecipeDao recipeDao) {
+        this.recipesFileName = recipesName;
```
Comment (evoevodin) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/PredefinedRecipeLoader.java]: Is it just a name or _path_ ?

Comment (akorneta) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/PredefinedRecipeLoader.java]: just a name

**Comment by garagatyi** [comment]
LGTM

**Pull request merged**